⬆️ 🎨 Allow the use of gtdb taxonomy in Autometa

Sidduppal opened the pull request:
- ⬆️ 🎨 Create `autometa-setup-gtdb` entrypoint for creating compatible GTDB databases
- ⬆️ `gtdb_to_taxdump` is now installed 
- ⬆️ 🎨 Put common functionality for parsing of taxonomic databases in `TaxonomyDatabase` class

Commands to replicate:

## Test GTDB

### Running `autometa-setup-gtdb` entrypoint

```
autometa-setup-gtdb --taxa-files /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/taxonomy_gtdb_releases/R207/bac120_taxonomy_r207.tsv /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/taxonomy_gtdb_releases/R207/ar53_taxonomy_r207.tsv --faa-tarball /media/bigdrive1/Databases/gtdb_proteins_aa_reps_r207.tar.gz --dbdir $HOME/gtdbTest --keep-temp
```

### Running `autometa-taxonomy-lca` entrypoint

```
autometa-taxonomy-lca --blast /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/78Mbp/78Mbp_gtdb_blastP.tsv --dbdir /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/gtdb_to_diamond --dbtype gtdb --sseqid2taxid-output 78Mbp_sseqid2taxid_test.tsv --lca-error-taxids 78Mbp_lcaErrorTaxids_test.tsv --verbose --lca-output 78Mbp_LCAout_test.tsv
```

### Running `autometa-taxonomy-majority-vote` entrypoint

```
autometa-taxonomy-majority-vote --lca 78Mbp_LCAout_test.tsv --output 78Mbp_gtdb_majority_vote.tsv --dbdir /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/gtdb_to_diamond --verbose --dbtype gtdb
```

### Running `autometa-taxonomy` entrypoint

```
autometa-taxonomy --votes 78Mbp_gtdb_majority_vote.tsv --assembly /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.filtered.fna --output testTaxonomy --split-rank-and-write superkingdom --dbdir /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/gtdb_to_diamond --dbtype gtdb
```

### Running `autometa-summary` entrypoint

```
autometa-binning-summary --binning-main /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.main.tsv --markers /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.markers.tsv --dbdir /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/gtdb_to_diamond --dbtype gtdb --output-stats binningSummartStats.tsv --output-taxonomy binningTaxa.tsv --output-metabins metaBins --metagenome /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.filtered.fna
```

## Test NCBI

### Running `autometa-taxonomy-lca` entrypoint

```
autometa-taxonomy-lca --blast /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.blastp.tsv --dbdir /media/bigdrive1/Databases/autometa_databases/ --lca-output lcaOut_ncbi.tsv --sseqid2taxid-output sseqid2taxid_ncbiTest.tsv --lca-error-taxids lcaError_ncbi.tsv --verbose
```

### Running `autometa-taxonomy-majority-vote` entrypoint

```
autometa-taxonomy-majority-vote --lca lcaOut_ncbi.tsv --output ncbi_majority_vote.tsv --dbdir /media/bigdrive1/Databases/autometa_databases/ --verbose --dbtype ncbi
```

### Running `autometa-taxonomy` entrypoint

```
autometa-taxonomy --votes ncbi_majority_vote.tsv --assembly /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.filtered.fna --output testNCBI --split-rank-and-write superkingdom --dbdir /media/bigdrive1/Databases/autometa_databases/
```

### Running `autometa-summary` entrypoint

```
autometa-binning-summary --binning-main /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.main.tsv --markers /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.markers.tsv --metagenome /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.filtered.fna --dbdir  /media/bigdrive1/Databases/autometa_databases/ --output-stats binningSummartStats2.tsv --output-taxonomy binningTaxa2.tsv --output-metabins metaBins2
```

## TODO:

- [ ] Update documentation. eg. Add paths to GTDB database files.
- [x] Add docstrings for `gtdb.py`


## PR checklist

- [X] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [X] Have you followed the pipeline conventions in the [contribution docs](https://github.com/KwanLab/autometa/tree/main/.github/CONTRIBUTING.md)

WiscEvan commented on .github/workflows/pytest_codecov.yml:
```diff
@@ -26,7 +26,7 @@ jobs:
       matrix:
         os: [ubuntu-latest]
-        python-version: [3.7]
+        python-version: [3.8]
```
Why is the python version for tests being changed here? Is this necessary for `gtdb_to_taxdump` installation?

Sidduppal replied:
It was giving an error with `3.7`, [link](https://github.com/KwanLab/Autometa/runs/7440617864?check_suite_focus=true#step:9:28). `from typing import Union, List, Literal` is only supported in versions >=3.8

WiscEvan commented on autometa-env.yml:
```diff
@@ -28,3 +28,5 @@ dependencies:
   - trimap
   - tsne
   - umap-learn>=0.5 # https://github.com/lmcinnes/umap/releases/tag/0.5.0 : umap, densmap
+  - pip:
+    - gtdb_to_taxdump # https://github.com/nick-youngblut/gtdb_to_taxdump
```
Database download and formatting appears to be working. Just need `gtdb_to_taxdump` to be available on bioconda then we can update the Autometa recipe and get this merged in

With `gtdb_to_taxdump` (v 0.1.8):

## Install this branch (and environment)

```bash
# In root of Autometa repo
# update env
mamba env update -n autometa -f=autometa-env.yml
# install autometa comands
make clean && make install
```

## Set config to point to where we want formatted GTDB database files

```bash
autometa-config \
    --section databases \
    --option gtdb \
    --value $HOME/autometa-testing/databases/gtdb
```

## Download/format GTDB databases to configured path

> NOTE: `$HOME` is not actually emitted. I've replaced the actual path with `$HOME` for security reasons.

```bash
(autometa) evan@userserver:~/Autometa$ autometa-update-databases --update-gtdb
[08/04/2022 05:10:33 PM DEBUG] autometa.taxonomy.gtdb: Running gtdb_to_taxdump.py: gtdb_to_taxdump.py https://data.gtdb.ecogenomic.org/releases/latest/ar53_taxonomy.tsv.gz https://data.gtdb.ecogenomic.org/releases/latest/bac120_taxonomy.tsv.gz --outdir $HOME/autometa-testing/databases/gtdb
2022-08-04 17:10:34,175 - Loading: https://data.gtdb.ecogenomic.org/releases/latest/ar53_taxonomy.tsv.gz
2022-08-04 17:10:35,548 - Loading: https://data.gtdb.ecogenomic.org/releases/latest/bac120_taxonomy.tsv.gz
2022-08-04 17:10:44,434 - File written: $HOME/autometa-testing/databases/gtdb/names.dmp
2022-08-04 17:10:44,434 - File written: $HOME/autometa-testing/databases/gtdb/nodes.dmp
2022-08-04 17:10:45,451 - File written: $HOME/autometa-testing/databases/gtdb/delnodes.dmp
2022-08-04 17:10:45,452 - File written: $HOME/autometa-testing/databases/gtdb/merged.dmp
[08/04/2022 05:10:45 PM DEBUG] autometa.config.databases: starting GTDB databases download
[08/04/2022 05:10:45 PM DEBUG] autometa.config.databases: wget https://data.gtdb.ecogenomic.org/releases/latest/genomic_files_reps/gtdb_proteins_aa_reps.tar.gz -O $HOME/autometa-testing/databases/gtdb/gtdb_proteins_aa_reps.tar.gz
[08/04/2022 06:11:46 PM DEBUG] autometa.taxonomy.gtdb: Running gtdb_to_diamond.py: gtdb_to_diamond.py $HOME/autometa-testing/databases/gtdb/gtdb_proteins_aa_reps.tar.gz $HOME/autometa-testing/databases/gtdb/names.dmp $HOME/autometa-testing/databases/gtdb/nodes.dmp --outdir $HOME/autometa-testing/databases/gtdb/gtdb_to_diamond
2022-08-04 18:11:46,417 - Read nodes.dmp file: $HOME/autometa-testing/databases/gtdb/nodes.dmp
2022-08-04 18:11:46,500 - File written: $HOME/autometa-testing/databases/gtdb/gtdb_to_diamond/nodes.dmp
2022-08-04 18:11:46,500 - Reading dumpfile: $HOME/autometa-testing/databases/gtdb/names.dmp
2022-08-04 18:11:47,561 -   File written: $HOME/autometa-testing/databases/gtdb/gtdb_to_diamond/names.dmp
2022-08-04 18:11:47,561 -   No. of accession<=>taxID pairs: 401808
2022-08-04 18:11:47,561 - Extracting tarball: $HOME/autometa-testing/databases/gtdb/gtdb_proteins_aa_reps.tar.gz
2022-08-04 18:11:47,561 -   Extracting to: gtdb_to_diamond_TMP
2022-08-04 18:24:08,234 -   No. of .faa(.gz) files: 65703
2022-08-04 18:24:08,256 - Creating accession2taxid table...
2022-08-04 18:24:08,376 -   File written: $HOME/autometa-testing/databases/gtdb/gtdb_to_diamond/accession2taxid.tsv
2022-08-04 18:24:08,376 - Formating & merging faa files...
2022-08-04 18:44:45,876 -   File written: $HOME/autometa-testing/databases/gtdb/gtdb_to_diamond/gtdb_all.faa
2022-08-04 18:44:45,876 -   No. of seqs. written: 200505361
2022-08-04 18:44:51,908 - Temp-dir removed: gtdb_to_diamond_TMP
[08/04/2022 06:44:51 PM DEBUG] autometa.common.external.diamond: diamond makedb --in $HOME/autometa-testing/databases/gtdb/gtdb_to_diamond/gtdb_all.faa --db $HOME/autometa-testing/databases/gtdb/gtdb_to_diamond/gtdb_all.dmnd -p 96
(autometa) evan@userserver:~/Autometa$
```

WiscEvan commented:
Looking at submitting a bioconda recipe for [gtdb_to_taxdump](https://github.com/nick-youngblut/gtdb_to_taxdump) and in the README it suggests using TaxonKit for retrieving stable taxids across GTDB releases (rather than arbitrarily assigning them as noted). Below is a screenshot from [the gtdb_to_taxdump summary section](https://github.com/nick-youngblut/gtdb_to_taxdump#summary):

<img width="855" alt="Screen Shot 2022-08-09 at 09 34 30" src="https://user-images.githubusercontent.com/25933122/183661714-bba0c434-b11e-47c3-a730-455e9dc91023.png">

-----

## Next Steps Towards Stability

TaxonKit is available via bioconda so would be easy to include in `autometa-env.yml` and documentation for generating these taxdump files from the GTDB looks straight-forward as outlined in the [GTDB-taxdump repo](https://github.com/shenwei356/gtdb-taxdump)

TaxonKit conda page: https://anaconda.org/bioconda/taxonkit
GTDB-taxdump page (using TaxonKit): https://github.com/shenwei356/gtdb-taxdump

NOTE: TaxonKit v0.12 or greater should be used. [(ref)](https://github.com/shenwei356/gtdb-taxdump#data-and-tools)

The taxdump files may also be downloaded directly from the releases page: https://github.com/shenwei356/gtdb-taxdump/releases which reduces the compute requirements.

That being said, if we would like to generate the GTDB taxdump files using this, the steps are outlined here: https://github.com/shenwei356/gtdb-taxdump#steps

Looking in to the future, this is probably the more appropriate route as taxids may be relied up across future GTDB releases.

> P.S there is also a related project linked (https://github.com/shenwei356/ictv-taxdump) that generates an NCBI taxdump for viruses and may be useful in the future (@kaw97, is this of any interest to you?)

chasemc commented:
Not sure if this affects what y'all are doing:
https://github.com/shenwei356/gtdb-taxdump/issues/2

shenwei356 commented:
> Not sure if this affects what y'all are doing: [shenwei356/gtdb-taxdump#2](https://github.com/shenwei356/gtdb-taxdump/issues/2)

Only `merged.dmp` and `delnodes.dmp` are affected. If you just need to use the taxonomy data of the current version, say R207, don't worry.

chasemc commented:
Thanks @shenwei356

Sidduppal commented:
@WiscEvan The scripts have been updated to use taxonkit's already generated files. I decided to use `merged.dmp` and `delnodes.dmp` as they were provided with the taxdump. Although if you think the issue mentioned above by @chasemc would skew the results a LOT we can remove their usage. I'm mentioning the commands for testing:

# Setting up the database

- The taxdump files needs to downloaded and extracted - [link](https://github.com/shenwei356/gtdb-taxdump/releases/download/v0.1.1/gtdb-taxdump.tar.gz)

```
autometa-config     \
--section databases     \
--option gtdb     \
--value /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/taxonKit/gtdb-taxdump/R207/test
autometa-update-databases --update-gtdb
```
# Run LCA

```
autometa-taxonomy-lca \
--blast 78mbp_metagenome.blastp.gtdb.tsv --dbdir /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/taxonKit/gtdb-taxdump/R207/test \
--dbtype gtdb \
--sseqid2taxid-output 78Mbp_sseqid2taxid_test.tsv \
--lca-error-taxids 78Mbp_lcaErrorTaxids_test.tsv \
--verbose \
--lca-output 78Mbp_LCAout_test.tsv
```
# Run Majority vote

```
autometa-taxonomy-majority-vote \
--lca 78Mbp_LCAout_test.tsv \
--output 78Mbp_gtdb_majority_vote.tsv \
--dbdir /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/taxonKit/gtdb-taxdump/R207/test \ 
--verbose \
--dbtype gtdb
```

# Run taxonomy

```
autometa-taxonomy \
--votes 78Mbp_gtdb_majority_vote.tsv \
--assembly /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.filtered.fna \
--output testTaxonomy \
--split-rank-and-write superkingdom \
--dbdir /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/taxonKit/gtdb-taxdump/R207/test \
--dbtype gtdb
```

# Run summary

```
autometa-binning-summary \
--binning-main 78_binningMain_gtdb.tsv \
--markers /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.markers.tsv --dbdir /media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/taxonKit/gtdb-taxdump/R207/test \
--dbtype gtdb \
--output-stats binningSummartStats.tsv --output-taxonomy binningTaxa.tsv \
--output-metabins metaBins \
--metagenome /media/bigdrive1/sidd/nextflow_trial/autometa_runs/78mbp_manual/interim/78mbp_metagenome.filtered.fna
```

By using this we don't need to add any additional dependency as well. Let me know what you think.

WiscEvan commented on autometa/config/default.config:
```diff
@@ -60,6 +61,9 @@ bacteria_single_copy = https://${markers:host}/KwanLab/Autometa/main/autometa/da
 archaea_single_copy = https://${markers:host}/KwanLab/Autometa/main/autometa/databases/markers/archaea.single_copy.hmm
 archaea_single_copy_cutoffs = https://${markers:host}/KwanLab/Autometa/main/autometa/databases/markers/archaea.single_copy.cutoffs
+proteins_aa_reps = https://${gtdb:host}/releases/${gtdb:release}/genomic_files_reps/gtdb_proteins_aa_reps.tar.gz
+gtdb_taxdmp = https://${markers:host}/shenwei356/gtdb-taxdump/releases/latest/download/gtdb-taxdump.tar.gz
```
This probably should not be referencing markers. Kind of confusing. The full Github path should probably be used. Also, this should either specify `taxonkit` or be generalized by some other selection to use `taxonkit` or another tool

WiscEvan replied:
```suggestion
taxonkit_gtdb_taxdump = https://raw.githubusercontent.com/shenwei356/gtdb-taxdump/releases/latest/download/gtdb-taxdump.tar.gz
```

WiscEvan commented on autometa/config/databases.py:
```diff
@@ -401,6 +403,30 @@ def download_ncbi_files(self, options: Iterable) -> None:
 
+    def download_gtdb_files(self) -> None:
+        proteins_aa_reps_url = self.config.get("database_urls", "proteins_aa_reps")
+        gtdb_taxdmp_url = self.config.get("database_urls", "gtdb_taxdmp")
```
```suggestion
        gtdb_taxdump_url = self.config.get("database_urls", "taxonkit_gtdb_taxdump")
```

WiscEvan commented on autometa/config/databases.py:
```diff
@@ -401,6 +403,30 @@ def download_ncbi_files(self, options: Iterable) -> None:
+
+        # User path:
+        proteins_aa_reps_filepath = self.config.get("gtdb", "proteins_aa_reps")
+        gtdb_taxdmp_filepath = self.config.get("gtdb", "gtdb_taxdmp")
+
+        urls = [proteins_aa_reps_url, gtdb_taxdmp_url]
```
```suggestion
        gtdb_taxdump_filepath = self.config.get("gtdb", "taxonkit_gtdb_taxdump")

        urls = [proteins_aa_reps_url, gtdb_taxdump_url]
```

WiscEvan commented on autometa/config/databases.py:
```diff
@@ -401,6 +403,30 @@ def download_ncbi_files(self, options: Iterable) -> None:
+        gtdb_taxdmp_filepath = self.config.get("gtdb", "gtdb_taxdmp")
+
+        urls = [proteins_aa_reps_url, gtdb_taxdmp_url]
+        filepaths = [proteins_aa_reps_filepath, gtdb_taxdmp_filepath]
```
```suggestion
        filepaths = [proteins_aa_reps_filepath, gtdb_taxdump_filepath]
```

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,362 @@
+        return test_f.read(2) == b"\x1f\x8b"
+
+
+def create_gtdb_db(reps_faa: str, dbdir: str):
```
✏️ :memo: This needs a docstring

WiscEvan replied:
And a typehint

```suggestion
def create_gtdb_db(reps_faa: str, dbdir: str) -> str:
```

WiscEvan replied:
It is not clear whether `reps_faa` is supposed to be a file path, directory path or tarball. This usage (and variable name) seems too ambiguous

Sidduppal replied:
It can be a tarball or a directory path.

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,362 @@
+            if file.endswith("_protein.faa"):
+                faa_file = os.path.join(dirpath, file)
+                # Regex to get the genome accession from the file path
+                acc = re.search(r"GCA_\d+.\d?|GCF_\d+.\d?", file)
```
This will cause an `AttributeError` if the regex does not match... Logic should be introduced to account for this.

Example error:

```python
In [7]: re.search(r"GCA_\d+.\d?|GCF_\d+.\d?", "some_incorrect_genome_file_path.fna.gz").group()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [1], in <module>
----> 1 re.search(r"GCA_\d+.\d?|GCF_\d+.\d?", "some_incorrect_genome_file_path.fna.gz").group()

AttributeError: 'NoneType' object has no attribute 'group'
```

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,362 @@
+        reps_faa = dbdir
+
+    faa_index = {}
+    for dirpath, dirs, files in os.walk(reps_faa):
+        for file in files:
+            if file.endswith("_protein.faa"):
+                faa_file = os.path.join(dirpath, file)
+                # Regex to get the genome accession from the file path
```
Rather than walking with multiple for loops, why not just glob `_protein.faa`?

`genome_protein_faa_filepaths = glob.glob(os.path.join(reps_faa, '*_protein.faa'), recursive=True)`


```python
faa_index = {
    re.search(r"GCA_\d+.\d?|GCF_\d+.\d?", fpath).group() : fpath
    for fpath in glob.glob(os.path.join(reps_faa, '*_protein.faa'), recursive=True)
}
```

```python
from typing import Dict
genome_protein_faa_filepaths = glob.glob(os.path.join(reps_faa, '*_protein.faa'), recursive=True)
faa_index: Dict[str, str] = {}
for genome_protein_faa_filepath in genome_protein_faa_filepaths:
    genome_acc_search = re.search(r"GCA_\d+.\d?|GCF_\d+.\d?", genome_protein_faa_filepath)
    if not genome_acc_search:
        continue
    genome_accession = genome_acc_search.group()
    faa_index[genome_accession] = genome_protein_faa_filepath
```

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,362 @@
+    if not os.path.isdir(dbdir):
+        os.makedirs(dbdir)
+
+    logger.debug("Merging faa files.")
```
More context could be provided here.

```suggestion
    logger.debug(f"Merging {len(faa_index):,} faa files.")
```

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,362 @@
+            with open(faa_file) as f_in:
+                for line in f_in:
+                    if line.startswith(">"):
+                        line = ">" + acc + " " + line.lstrip(">")
```
An f-string may be more clear here

```suggestion
                        seqheader = line.lstrip(">")
                        line = f">{acc} {seqheader}"
```

WiscEvan reviewed (changes requested):
1. It does not appear that you are using `delnodes.dmp` in `gtdb.py` other than reading the file.
2. It does not appear that you are using `merged.dmp` in `gtdb.py` other than reading the file. 
3. Tests are currently failing and should be double-checked

There are some other comments attached. I'll review more after these are addressed. 👍 🚧 👷

WiscEvan reviewed (changes requested):
There are a few files that I've ignored for the time being. It looks like there are some changes that may not be functional. It looks like there are a few bugs that need to be addressed. 🚧 👷 ⚒️

Sidduppal commented on .github/workflows/pytest_codecov.yml:
```diff
@@ -26,7 +26,7 @@ jobs:
       matrix:
         os: [ubuntu-latest]
-        python-version: [3.7]
+        python-version: [3.8]
```
Yes, it's needed else one of the tests is failing.

Sidduppal commented on autometa-env.yml:
```diff
@@ -12,6 +12,7 @@ dependencies:
   - gdown
   - hdbscan
   - hmmer
+  - joblib==1.1.0
```
It's not relevant to GTDB, but some tests were failing if this is not set.

Sidduppal commented on autometa/binning/summary.py:
```diff
@@ -377,14 +390,17 @@ def main():
             )
         else:
-            ncbi = NCBI(dirpath=args.ncbi)
+            if args.dbtype == "ncbi":
+                taxa_db = NCBI(dbdir=args.dbdir)
+            elif args.dbtype == "gtdb":
+                taxa_db = GTDB(dbdir=args.dbdir)
```
At least for me the `if/else` is easier to read and understand, hence I'm leaving it that way for now. Let me know if you still want me to change it.

Sidduppal commented on autometa/taxonomy/database.py:
```diff
@@ -0,0 +1,350 @@
+logger = logging.getLogger(__name__)
+
+
+class TaxonomyDatabase(ABC):
```
@WiscEvan If I remember correctly this function/class was written by you and I don't think I fully understand its role. It will be great if you can add the docstring instead 😄

Sidduppal commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,393 @@
+                        line = f">{acc} {seqheader}"
+                    f_out.write(line)
+    logger.debug(f"Combined GTDB faa file written to {dbdir}")
+    return os.path.join(dbdir, "gtdb.faa")
```
I think both should be `gtdb.faa`. Thanks for catching this.

Sidduppal commented on autometa/taxonomy/lca.py:
```diff
@@ -774,7 +692,12 @@ def main():
     args = parser.parse_args()
 
-    lca = LCA(dbdir=args.dbdir, verbose=args.verbose, cache=args.cache)
+    if args.dbtype == "ncbi":
+        taxonomy_db = NCBI(args.dbdir)
+    elif args.dbtype == "gtdb":
+        taxonomy_db = GTDB(args.dbdir)
```
At least for me the if/else is easier to read and understand, hence I'm leaving it that way for now. Let me know if you still want me to change it.

Sidduppal commented on autometa/taxonomy/majority_vote.py:
```diff
@@ -312,10 +324,15 @@ def main():
     )
     args = parser.parse_args()
 
+    if args.dbtype == "ncbi":
+        taxa_db = NCBI(args.dbdir, verbose=args.verbose)
+    elif args.dbtype == "gtdb":
+        taxa_db = GTDB(args.dbdir, verbose=args.verbose)
```
At least for me the if/else is easier to read and understand, hence I'm leaving it that way for now. Let me know if you still want me to change it.

Sidduppal commented on workflows/autometa-gtdb.sh:
```diff
@@ -0,0 +1,298 @@
+        --binning-main $binning_main \
+        --markers $markers \
+        --metagenome $assembly \
+        --dbdir $gtdb \
+        --dbtype gtdb \
```
This workflow has not been implemented. I only worked on implementing the codebase and not the actual workflow.

Sidduppal commented on workflows/autometa-gtdb.sh:
```diff
@@ -0,0 +1,298 @@
+
+records = SeqIO.parse(${outdir}/${simpleName}.bacteria.fna, 'fasta')
+
+prodigal.orf_records_from_contigs(records, ${outdir}/${simpleName}.orfs.faa)
```
This workflow has not been implemented. I only worked on implementing the codebase and not the actual workflow.

Sidduppal replied:
I think we should have the workflow as a separate PR.

Sidduppal commented on autometa/taxonomy/vote.py:
```diff
@@ -95,6 +100,8 @@ def assign(
     prot_orfs = prot_orfs if prot_orfs else os.path.join(outdir, "orfs.faa")
     nucl_orfs = nucl_orfs if nucl_orfs else os.path.join(outdir, "orfs.fna")
 
+    taxa_db = NCBI(dbdir=dbdir) if dbtype == "ncbi" else GTDB(dbdir=dbdir)
```
> Dispatcher may be worth implementing here as well

At least for me the if/else is easier to read and understand, hence I'm leaving it that way for now. Let me know if you still want me to change it.

Sidduppal commented on autometa/taxonomy/vote.py:
```diff
@@ -339,13 +347,24 @@ def main():
+    )
     args = parser.parse_args()
 
+    if args.dbtype == "ncbi":
+        taxa_db = NCBI(args.dbdir)
+    elif args.dbtype == "gtdb":
+        taxa_db = GTDB(args.dbdir)
```
At least for me the if/else is easier to read and understand, hence I'm leaving it that way for now. Let me know if you still want me to change it.

Sidduppal commented on workflows/autometa-gtdb.sh:
```diff
@@ -0,0 +1,298 @@
+# script:
+autometa-taxonomy-lca \
+    --blast $blast \
+    --dbdir $ncbi \
+    --lca-output $lca \
+    --sseqid2taxid-output $sseqid2taxid \
+    --lca-error-taxids $errorTaxids
+
+# Step 4.2: Perform Modified Majority vote of ORF LCAs for all contigs that returned hits in blast search
+
+# input:
+# $lca --> Generated by step 4.1
+# $ncbi --> User Input
+
+# output:
+votes="${outdir}/${simpleName}.taxids.tsv"
+
+# script:
+autometa-taxonomy-majority-vote --lca $lca --output $votes --dbdir $ncbi
```
This workflow has not been implemented. I only worked on implementing the codebase and not the actual workflow.
I think we should have the workflow as a separate PR.

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,368 @@
+        genome_acc_search = re.search(
+            r"GCA_\d+.\d?|GCF_\d+.\d?", genome_protein_faa_filepath
+        )
+        if not genome_acc_search:
+            continue
+        faa_index[genome_protein_faa_filepath] = genome_acc_search.group()
```
This can be simplified by switching the boolean expression.

> In `re.search(pattern, file_string)` is the entire path returned from `genome_acc_search.group()`? or only the GCA or GCF string?

> Yep, looks like it simply returns the match

```python
genome_protein_faa_filepath = "gtdb_release207_v2_genomic_files_reps/genomic_files_reps/protein_faa_reps/bacteria/GB_GCA_018826885.1_protein.faa"

re.search(r"GCA_\d+.\d?|GCF_\d+.\d?", genome_protein_faa_filepath).group()
>>>'GCA_018826885.1'
```

```suggestion
        if genome_acc_search:
            faa_index[genome_protein_faa_filepath] = genome_acc_search.group()
```

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,368 @@
+                        line = f"\n>{acc} {seqheader}"
+                    f_out.write(line)
+    logger.debug(f"Combined GTDB faa file written to {dbdir}")
+    return os.path.join(dbdir, "gtdb.faa")
```
It might be better here to define the `"gtdb.faa"` filepath once and use the variable in both locations.

```suggestion
    return combined_faa
```

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,368 @@
+                        seqheader = line.lstrip(">")
+                        line = f"\n>{acc} {seqheader}"
+                    f_out.write(line)
+    logger.debug(f"Combined GTDB faa file written to {dbdir}")
```
Might as well log the actual filepath instead of just the directory.

```suggestion
    logger.debug(f"Combined GTDB faa file written to {combined_faa}")
```

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,368 @@
+    def parse_delnodes(self) -> Set[int]:
+        """
+        Parse the delnodes.dmp database
+        Note: This is performed when a new NCBI class instance is constructed
```
```suggestion
```

WiscEvan commented on autometa/taxonomy/gtdb.py:
```diff
@@ -0,0 +1,368 @@
+    def parse_merged(self) -> Dict[int, int]:
+        """
+        Parse the merged.dmp database
+        Note: This is performed when a new NCBI class instance is constructed
```
```suggestion
```

These docstrings (`parse_merged` and `parse_delnodes`) should be placed in the corresponding `TaxonomyDatabase` abstract methods, details specifically about GTDB may be placed here.

WiscEvan reviewed (changes requested):
Getting closer, some minor edits and some documentation still needed. 👍 🏗️

WiscEvan commented on autometa/taxonomy/database.py:
```diff
@@ -0,0 +1,406 @@
+        """
+
+    @abstractmethod
+    def parse_delnodes(self) -> Dict[int, int]:
```
```suggestion
    def parse_delnodes(self) -> Set[int]:
```

WiscEvan commented on autometa/taxonomy/database.py:
```diff
@@ -0,0 +1,406 @@
+
+    e.g.
+
+    class GTDB(TaxonomyDatabase):
+        def parse_nodes(self):
+            ...
+        def parse_nodes(self):
+            ...
+        def parse_merged(self):
+            ...
+        def parse_delnodes(self):
+            ...
+        def convert_accessions_to_taxids(self):
+            ...
```
```suggestion
    class GTDB(TaxonomyDatabase):
        def __init__(self, ...):
            self.nodes = self.parse_nodes()
            self.names = self.parse_names()            
            self.merged = self.parse_merged()
            self.delnodes = self.parse_delnodes()
            ...
        def parse_nodes(self):
            ...
        def parse_nodes(self):
            ...
        def parse_merged(self):
            ...
        def parse_delnodes(self):
            ...
        def convert_accessions_to_taxids(self):
            ...
```

WiscEvan commented:
The tests and CI/CD still need to be resolved, but I think we're almost there.

WiscEvan commented on docs/source/databases.rst:
```diff
@@ -45,3 +46,57 @@ The previous command will download the following NCBI databases:
+
+Once unzipped `gtdb-taxdump.tar.gz` will have the taxdump files of all the respective GTDB releases. Make sure that the release you use is in line with the `gtdb_proteins_aa_reps.tar.gz` release version. It's better to always use the latest version. 
+
+All the taxonomy files for a specific taxonomy database should be in a single directory. You can now copy the taxdump files of the desired release version in the sample directory as `gtdb.dmnd`
```
I'm not sure I follow this statement.

What does this mean?

>You can now copy the taxdump files of the desired release version in the sample directory as `gtdb.dmnd`

WiscEvan commented on docs/source/databases.rst:
```diff
@@ -45,3 +46,57 @@ The previous command will download the following NCBI databases:
+
+.. code-block:: bash
+
+    python -m autometa.taxonomy.gtdb --reps-faa <path/to/gtdb_proteins_aa_reps.tar.gz> --dbdir <path/to/output_directory> --cpus 20
```
Is there an entrypoint in `setup.py` to run this as a module? The `python -m autometa.taxonomy.gtdb` command may only work if the user has installed from source.

WiscEvan commented on docs/source/databases.rst:
```diff
@@ -45,3 +46,57 @@ The previous command will download the following NCBI databases:
+
+.. note::
+
+    Again Make sure that the formatted `gtdb_proteins_aa_reps.tar.gz` databse and gtdb taxdump files are in the same directory. 
```
Is the 'formatted `gtdb_proteins_aa_reps.tar.gz` database `gtdb.faa`? 

```suggestion
    The formatted `gtdb_proteins_aa_reps.tar.gz` database and GTDB taxdump files should be in the same directory. 
```

WiscEvan commented on docs/source/databases.rst:
```diff
@@ -45,3 +46,57 @@ The previous command will download the following NCBI databases:
+    - `gtdb-taxdump.tar.gz <https://github.com/shenwei356/gtdb-taxdump/releases/latest/download/gtdb-taxdump.tar.gz>`_
+
+
+Once unzipped `gtdb-taxdump.tar.gz` will have the taxdump files of all the respective GTDB releases. Make sure that the release you use is in line with the `gtdb_proteins_aa_reps.tar.gz` release version. It's better to always use the latest version. 
```
This section is not clear and you should break this in to multiple lines.

Where does `gtdb.faa` come in? When is `gtdb.dmnd` generated, if at all?

WiscEvan commented on workflows/autometa_gtdb_flagged.sh:
```diff
@@ -0,0 +1,538 @@
+   echo "-o     Directory path to where the output will go"
+   echo "-a     File path to your assembly file (include file name)"
+   echo "-s     Simple name to prefix output files"
+   echo "-n     Path to NCBI databases directory"
+   echo "-g     Path to GTDB databases directory"
```
The `autometa_flagged.sh` workflow should also be changed to comply with the `--dbdir` and `--dbtype` changes.

WiscEvan reviewed (changes requested):
🚧 👷 🏗️ Almost there, just a few adjustments and we should be able to finalize this integration.

1. :fire::green_heart: The test workflows should be removed. These are not following DRY (don't repeat yourself) principles. These are simply testing the test scripts. If adding tests, the workflows we expect users to download should be tested. 
2. 🐛 🛠️ I'm not sure how you encountered the error below. Sam and I implemented a `set -x` / `{ set +x; } 2>/dev/null` routine before and after running each module (https://github.com/KwanLab/Autometa/blob/gtdb_to_autometa/workflows/autometa_flagged.sh) that should allow easier inspection of a user's parameter configurations without them sending the entire submit file.

```bash
/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/umap/umap_.py:2344: UserWarning: n_neighbors is larger than the dataset size; truncating to X.shape[0] - 1
  warn(
Traceback (most recent call last):
  File "/home/sidd/miniconda3/envs/autometa_aims/bin/autometa-binning-ldm", line 33, in <module>
    sys.exit(load_entry_point('Autometa==2.1.0', 'console_scripts', 'autometa-binning-ldm')())
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/Autometa-2.1.0-py3.9.egg/autometa/binning/large_data_mode.py", line 831, in main
    main_out = cluster_by_taxon_partitioning(
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/Autometa-2.1.0-py3.9.egg/autometa/binning/large_data_mode.py", line 441, in cluster_by_taxon_partitioning
    rank_embedding = get_kmer_embedding(
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/Autometa-2.1.0-py3.9.egg/autometa/binning/large_data_mode.py", line 112, in get_kmer_embedding
    embedding.to_csv(cache_fpath, sep="\t", index=True, header=True)
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/core/generic.py", line 3551, in to_csv
    return DataFrameRenderer(formatter).to_csv(
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/io/formats/format.py", line 1180, in to_csv
    csv_formatter.save()
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/io/formats/csvs.py", line 241, in save
    with get_handle(
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/io/common.py", line 694, in get_handle
    check_parent_directory(str(handle))
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/io/common.py", line 568, in check_parent_directory
    raise OSError(rf"Cannot save file into a non-existent directory: '{parent}'")
OSError: Cannot save file into a non-existent directory: '78mbp_metagenome_Autometa_Output3/78mbp_metagenome_bacteria_cache/species'
```

3. 📝 I've placed some comments regarding documentation that still needs to be addressed.
4. The GTDB workflows require changes to some of the logic to reduce redundancy (and also contain some minor nitpicks for code transparency).

Sidduppal commented:
Addressed all the comments. 

> 🐛 🛠️ I'm not sure how you encountered the error below. Sam and I implemented a set -x / { set +x; } 2>/dev/null routine before and after running each module (https://github.com/KwanLab/Autometa/blob/gtdb_to_autometa/workflows/autometa_flagged.sh) that should allow easier inspection of a user's parameter configurations without them sending the entire submit file.


Still getting the following error when running [autometa-large-data-mode-gtdb.sh](https://github.com/KwanLab/Autometa/blob/gtdb_to_autometa/workflows/autometa-large-data-mode-gtdb.sh). I'm getting the error during the large-data-mode binning ([lines](https://github.com/KwanLab/Autometa/blob/gtdb_to_autometa/workflows/autometa-large-data-mode-gtdb.sh#L308-L329)).

The set -x / { set +x; } 2>/dev/null routine has only been set for `autometa_flagged.sh` and no other script. I'm running the large-data-mode workflow which does not have the above stated routine. The bug seems to be with the large-data-mode implementation as all the other workflows are running fine.


```
[10/16/2022 04:41:27 PM DEBUG] autometa.common.kmers: umap: 10 data points and 10 dimensions
[10/16/2022 04:41:27 PM DEBUG] autometa.common.kmers: Performing embedding with umap (seed 42)
/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/umap/umap_.py:2344: UserWarning: n_neighbors is larger than the dataset size; truncating to X.shape[0] - 1
  warn(
Traceback (most recent call last):
  File "/home/sidd/miniconda3/envs/autometa_aims/bin/autometa-binning-ldm", line 33, in <module>
    sys.exit(load_entry_point('Autometa==2.1.0', 'console_scripts', 'autometa-binning-ldm')())
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/Autometa-2.1.0-py3.9.egg/autometa/binning/large_data_mode.py", line 831, in main
    main_out = cluster_by_taxon_partitioning(
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/Autometa-2.1.0-py3.9.egg/autometa/binning/large_data_mode.py", line 441, in cluster_by_taxon_partitioning
    rank_embedding = get_kmer_embedding(
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/Autometa-2.1.0-py3.9.egg/autometa/binning/large_data_mode.py", line 112, in get_kmer_embedding
    embedding.to_csv(cache_fpath, sep="\t", index=True, header=True)
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/core/generic.py", line 3551, in to_csv
    return DataFrameRenderer(formatter).to_csv(
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/io/formats/format.py", line 1180, in to_csv
    csv_formatter.save()
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/io/formats/csvs.py", line 241, in save
    with get_handle(
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/io/common.py", line 694, in get_handle
    check_parent_directory(str(handle))
  File "/home/sidd/miniconda3/envs/autometa_aims/lib/python3.9/site-packages/pandas/io/common.py", line 568, in check_parent_directory
    raise OSError(rf"Cannot save file into a non-existent directory: '{parent}'")
OSError: Cannot save file into a non-existent directory: '/media/bigdrive1/sidd/autometa_aim1_1/data/external/gtdbData_r207v2/test1/taxonkit/78mbp_metagenome_Autometa_Output3/78mbp_metagenome_bacteria_cache/species'
```

WiscEvan reviewed (changes requested):
### Addition of GTDB sub-workflow as an option to existing workflows

Instead of multiple workflow files which is creating more repetition and making code review more cumbersome, why not add GTDB as a subworkflow in the existing workflow files? I.e. the ability to specify the use of the GTDB subworkflow in `autometa_flagged.sh`, etc. This would reduce the amount of code as well as minimize the opportunities for introducing bugs. This should also make using Autometa more approachable/less intimidating. We do not want to overwhelm the user with a multitude of workflow files (i.e. choices) when most of the processes between the files are the same.

### 🔥 Remove `workflowTest.sh`

Any tests added should be run against the autometa workflows, not a copy/pasted version but using the autometa workflow files in the repo.

### 🎨 🔍 Addition of `set -x` / `{ set +x; } 2>/dev/null` routine

> 🐛 🛠️ I'm not sure how you encountered the error below. Sam and I implemented a `set -x` / `{ set +x; } 2>/dev/null` routine before and after running each module (https://github.com/KwanLab/Autometa/blob/gtdb_to_autometa/workflows/autometa_flagged.sh) that should allow easier inspection of a user's parameter configurations without them sending the entire submit file.

I mentioned this as a routine to add to the workflow. This would emit the command being run and therefore would skip the additional step of asking for the submission parameters (they would be emitted in the log). So just to clarify, I do not suspect this routine is a cause of the issue encountered but rather would help to troubleshoot the issue.


> The set -x / { set +x; } 2>/dev/null routine has only been set for `autometa_flagged.sh` and no other script. I'm running the large-data-mode workflow which does not have the above stated routine. The bug seems to be with the large-data-mode implementation as all the other workflows are running fine.

###

WiscEvan reviewed (changes requested):
It looks like the GTDB taxon assignment sub-workflow requires a few additional checks and some comments linking to documentation

WiscEvan commented on autometa/binning/large_data_mode.py:
```diff
@@ -305,18 +305,25 @@ def cluster_by_taxon_partitioning(
     FileNotFoundError
         Provided `binning_checkpoints_fpath` does not exist
+    TableFormatError
+        No marker information is availble for contigs to be binned.
```
```suggestion
        No marker information is available for contigs to be binned.
```

WiscEvan reviewed (changes requested):
Still not sure why you are running in to an error with the `autometa-large-data-mode.sh` workflow. From inspecting your output directory, it looks like only 11 contigs were classified as bacteria, so maybe this is causing the issue? Large data mode was not intended for 11 contigs

WiscEvan commented on workflows/autometa.sh:
```diff
@@ -145,20 +176,164 @@ autometa-taxonomy-majority-vote --lca $lca --output $votes --dbdir $ncbi
+    # $blast --> Generated from step 5.2
+    # $gtdb --> User Input
+    # $dbtype --> Updated according to $taxa_routine
+    dbtype="gtdb"
```
`dbtype="gtdb"` is also on line 247. Is this necessary, or can we delete this?

WiscEvan commented on workflows/autometa.sh:
```diff
@@ -145,20 +176,164 @@ autometa-taxonomy-majority-vote --lca $lca --output $votes --dbdir $ncbi
+
+    # output:
+    # Will write recovered superkingdoms to $outdir
+    # e.g. ${outdir}/${prefix}.gtdb.bacteria.fna
+    # e.g. ${outdir}/${prefix}.gtdb.archaea.fna
+    # e.g. ${outdir}/${prefix}.gtdb.taxonomy.tsv
```
`gtdb`, i.e. `$dbtype` is included in `$prefix`

```suggestion
    # e.g. ${outdir}/${prefix}.bacteria.fna
    # e.g. ${outdir}/${prefix}.archaea.fna
    # e.g. ${outdir}/${prefix}.taxonomy.tsv
```

WiscEvan commented on workflows/autometa.sh:
```diff
@@ -145,20 +176,162 @@ autometa-taxonomy-majority-vote --lca $lca --output $votes --dbdir $ncbi
+
+        kingdom_fasta="${outdir}/${prefix}.${kingdom}.fna"
+        
+        contig_ids="${outdir}/${prefix}.${kingdom}.contigIDs.txt"
```
This writes sequence headers with the `_` suffix. This should be more appropriately named `{...}.orfPrefixes.txt` and the variable name changed to `orf_id_prefixes`

```suggestion
        orf_prefixes="${outdir}/${prefix}.${kingdom}.orfPrefixes.txt"
```

NOTE: `contig_ids` will need to be changed to `orf_prefixes` where appropriate as well

WiscEvan commented on workflows/autometa.sh:
```diff
@@ -145,20 +176,162 @@ autometa-taxonomy-majority-vote --lca $lca --output $votes --dbdir $ncbi
+            sed 's/$/_/' | \
+            cut -f1 -d" " > $contig_ids
+        # Retrieve ORF IDs from contig IDs
+        grep -f $contig_ids $orfs | sed 's/^>//' | cut -f1 -d" " > $orf_ids
```
NOTE: See comment above

```suggestion
        grep -f $orf_prefixes $orfs | sed 's/^>//' | cut -f1 -d" " > $orf_ids
```

WiscEvan commented on workflows/autometa.sh:
```diff
@@ -145,20 +176,162 @@ autometa-taxonomy-majority-vote --lca $lca --output $votes --dbdir $ncbi
+    # Step 5.2: Run blastp
+    # input:
+    # $gtdb_input_orfs --> Generated from step 5.1
+    gtdb_dmnd_db="${gtdb}/gtdb.dmnd" # generated using autometa-setup-gtdb (Must be performed prior to using this script)
```
I ended up using find here for CHTC, which may be a little more robust. Do you think this is appropriate?

```suggestion
    gtdb_dmnd_db=$(find $gtdb -name "gtdb.dmnd") # generated using autometa-setup-gtdb (Must be performed prior to using this script)
```

Sidduppal replied:
Not a bad idea. I can do that.

WiscEvan reviewed (approved):
Looks good. 👍  Release forthcoming
